src/main/cpp/core/Atlas.h

```diff
@@ -7,35 +7,58 @@
  */
 
 #include <cstddef>
+#include <iostream>
 #include <map>
+#include <string>
 #include <utility>
 #include "core/ClusterType.h"
 #include "geo/GeoPosition.h"
 
 namespace stride {
-
 /**
- * A map from clusters, identified by their ID and type, to geopositions.
+ * A map storing geopositions for clusters, and town data for geopositions.
  */
 class Atlas
 {
 public:
-	using Key = std::pair<std::size_t, ClusterType>;
-	using Map = std::map<Key, geo::GeoPosition>;
+	struct Town
+	{
+		std::string name;
+		unsigned int size;
+		static std::size_t newId;
+		std::size_t id;
+
+		Town(const std::string& name, const unsigned int size) : name(name), size(size), id(newId++) {}
+	};
+
+	using ClusterKey = std::pair<std::size_t, ClusterType>;
+	using ClusterMap = std::map<ClusterKey, geo::GeoPosition>;
+	using TownMap = std::map<geo::GeoPosition, Town>;
 
 	Atlas() {}
-	Atlas(const Map& map) : map(map) {}
+	Atlas(const ClusterMap& cluster_map) : cluster_map(cluster_map) {}
 
 	/// Look up a cluster's position. If it isn't found, throw std::out_of_range.
-	const geo::GeoPosition& Lookup(const Key& key) const { return map.at(key); }
+	const geo::GeoPosition& LookupPosition(const ClusterKey& key) const { return cluster_map.at(key); }
 
-	/// Store a GeoPosition in the atlas.
-	auto Emplace(const Atlas::Key& key, const geo::GeoPosition& pos) -> decltype(Map().emplace(key, pos))
+	/// Look up a cluster's town. If it isn't found, throw std::out_of_range.
+	const Town LookupTown(const ClusterKey& key) const { return town_map.at(cluster_map.at(key)); }
```

Comment:
Did you mean `const Town&`? A return value of `const Town` is identical to a return value of `Town` in every way, except for the added visual noise.